### PR TITLE
Fix the Spark 340 build error related to `mapKeyNotExistError`

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>
Related to #6383 

## Rational of this PR
`mapKeyNotExistError` has been removed since spark 3.4.0. We update the shim layers to match the change.
Please look at #6383 to find more context.

#### branch 22.12
```
...
[ERROR] [Error] /home/remziy/working/rapids/spark-rapids/sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:38: value mapKeyNotExistError is not a member of object org.apache.spark.sql.errors.QueryExecutionErrors
...
[ERROR] 35 errors found
```

#### this patch
```
...
[ERROR] 34 errors found
```

## Changes in the PR
1. Move the `RapidsErrorUtils.scala` under `spark330+` to `spark330until340`
2. Add a new `RapidsErrorUtils.scala` for Spark 340, in which we
  - mark the function `mapKeyNotExistError` as deprecated
  - change the the implementation of `mapKeyNotExistError` to always throw `UnsupportedOperationException` at runtime.
  In this way, we can make sure that developers who occasionally use `mapKeyNotExistError` in their code will get warnings at both compile time and runtime when building Spark 340.

### CODE REVIEW

File: sql-plugin/src/main/340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  @deprecated("this function has been removed", "Spark 3.4.0")
+  def mapKeyNotExistError(
+      key: String,
+      keyType: DataType,
+      origin: Origin): NoSuchElementException = {
+    throw new UnsupportedOperationException(
+      "`mapKeyNotExistError` has been removed since Spark 3.4.0. ")
+  }
```
Comment (HaoYang670) [inline sql-plugin/src/main/340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:41]: Changes are here

**Comment by HaoYang670** [comment]
@nartal1 Could you please help to review?

File: sql-plugin/src/main/340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
+/*
```
Comment (tgravescs) [inline sql-plugin/src/main/340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:1]: all of the files for 3.4.0 should go into a 340+ directory with the assumption that they will work for other versions like 3.4.1.

File: sql-plugin/src/main/340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  def invalidArrayIndexError(index: Int, numElements: Int,
+      isElementAtF: Boolean = false): ArrayIndexOutOfBoundsException = {
+    if (isElementAtF) {
+      QueryExecutionErrors.invalidElementAtIndexError(index, numElements)
+    } else {
+      QueryExecutionErrors.invalidArrayIndexError(index, numElements)
+    }
+  }
+
+  @deprecated("this function has been removed", "Spark 3.4.0")
```
Comment (tgravescs) [inline sql-plugin/src/main/340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:34]: this doesn't make sense to me, the thing removed was QueryExecutionErrors.mapKeyNotExistsError, not  the RapdisErrorUtils.mapKeyNotExistsError.  just remove this @deprecated.

**Review by tgravescs — COMMENTED** [review]
so this is fine to fix the compile error, but we didn't change to match the behavior. I assume that is why you left the issue open so that is good.

**Comment by nartal1** [comment]
While working on other build issues, I had similar solution to move RapidsErrorUtils.scala from `330+` to `spark330until340` but it didn't look ideal because we end up copying same methods in `spark330until340` shim  and `spark340+` including those whose behavior doesn't change between Spark versions. 
Met with @gerashegalov and his suggestion is to create a `trait RapidsErrorUtilsFor33Xand34X` in `330+` shim which has shared methods. And have `object RapidsErrorUtils` extend this from `spark330until340` and `340+` which will have version specific methods. I am working on that in my current implementation.

**Comment by HaoYang670** [comment]
Convert to draft because the implementation in #6876 is better.

**Comment by HaoYang670** [comment]
I will close this PR and create a new one because there are so many conflicts

**Pull request closed without merging**